### PR TITLE
Probes dma fix

### PR DESCRIPTION
**Pull request by jsarha**
Couple of fixes for probes feature.

### CODE REVIEW

**Review by lgirdwood — APPROVED** [review]
@jsarha what about Zephyr DMA integration ? Is this a next step after IPC4 ?

**Comment by jsarha** [comment]
> @jsarha what about Zephyr DMA integration ? Is this a next step after IPC4 ?

This is just to get the old feature working and ready for testing in CI. I have not given any thought to IPC4 probes implementation on FW side yet. The next step is to get IPC4 probes working on Linux side with closed source FW, but if that does not start working easily, I may have take on the IPC4 probes SOF FW implementation... but Zephyr, that is beyond my horizon ATM.

File: src/ipc/dma-copy.c [hunk 1]
```diff
@@ -74,10 +74,30 @@ int dma_copy_to_host(struct dma_copy *dc, struct dma_sg_config *host_sg,
...
+		return -EINVAL;
+
+	/* tell gateway to copy */
+	ret = dma_copy_legacy(dc->chan, size, 0);
+	if (ret < 0)
+		return ret;
+
+	/* bytes copied */
+	return size;
+}
```
Comment (ujfalusi) [inline src/ipc/dma-copy.c:93]: Would it make sense the have similar wrapping as with `!CONFIG_DMA_GW`
```c
static int
dma_copy_to_host_flags(struct dma_copy *dc, struct dma_sg_config *host_sg,
		       int32_t host_offset, void *local_ptr, int32_t size,
		       uint32_t flags)
{
	int ret;

	/* tell gateway to copy */
	ret = dma_copy_legacy(dc->chan, size, flags);
	if (ret < 0)
		return ret;

	/* bytes copied */
	return size;
}

int dma_copy_to_host(struct dma_copy *dc, struct dma_sg_config *host_sg,
		     int32_t host_offset, void *local_ptr, int32_t size)
{
	return dma_copy_to_host_flags(dc, host_sg, host_offset, local_ptr, size,
				      DMA_COPY_BLOCKING);
}

int dma_copy_to_host_nowait(struct dma_copy *dc, struct dma_sg_config *host_sg,
			    int32_t host_offset, void *local_ptr, int32_t size)
{
	return dma_copy_to_host_flags(dc, host_sg, host_offset, local_ptr, size, 0);
}
```
Btw, do you know why ONE_SHOT is not applicable for DMA_GW?

File: src/ipc/dma-copy.c [hunk 2]
```diff
@@ -74,10 +74,30 @@ int dma_copy_to_host(struct dma_copy *dc, struct dma_sg_config *host_sg,
...
+
 #else /* CONFIG_DMA_GW */
 
-int dma_copy_to_host(struct dma_copy *dc, struct dma_sg_config *host_sg,
-		     int32_t host_offset, void *local_ptr, int32_t size)
+static
+int dma_copy_to_host_flags(struct dma_copy *dc, struct dma_sg_config *host_sg,
```
Comment (ujfalusi) [inline src/ipc/dma-copy.c:98]: Usually we do the line split:
```c
static int
dma_copy_to_host_flags(struct dma_copy *dc, struct dma_sg_config *host_sg,
		       int32_t host_offset, void *local_ptr, int32_t size,
		       uint32_t flags)
```

File: src/ipc/dma-copy.c (hunk 1 above)
Comment (ujfalusi) [reply src/ipc/dma-copy.c:93]: If we ONE_SHOT is good for both DMAs then we could have only the `dma_copy_to_host_flags()` ifdefed.

**Review by ujfalusi — COMMENTED** [review]
@jsarha, thanks for the fix, it works great!

It would be really great to understand _why_ we need non blocking DMA and how that actually works for the probes and why it is bad for the dtrace.

@ranj063, @kv2019i, @lyakh, @juimonen, @singalsu, do you have insights?

File: src/ipc/dma-copy.c [hunk 3]
```diff
@@ -74,10 +74,30 @@ int dma_copy_to_host(struct dma_copy *dc, struct dma_sg_config *host_sg,
...
 }
 
+int dma_copy_to_host_nowait(struct dma_copy *dc, struct dma_sg_config *host_sg,
+			    int32_t host_offset, void *local_ptr, int32_t size)
+{
+	int ret;
+
+	/* return if DMA channel is not get yet */
+	if (!dc->chan)
+		return -EINVAL;
```
[the remark this answers is not in the document]
Comment (jsarha) [reply src/ipc/dma-copy.c:84]: Oh, I somehow managed copy-paste the function between 89ec377cb55de04270dd6fa2bd8672c3dea3e83d..a487ca9441fb5d8738565db377ef7d3ad85579ac. Will fix.

File: src/ipc/dma-copy.c (hunk 1 above)
Comment (jsarha) [reply src/ipc/dma-copy.c:93]: If I am not mistaken, BLOCKING is the only flag intel/hda-dma.c is sensitive to. But I have no idea about the other dma controllers. You tell me, should I dare to drop ONE_SHOT to make the code tiny bit cleaner? When I last did that I broke the probes :laughing: . Then again there was no non blocking version for !CONFIG_DMA_GW despite the function  being called *_nowait()...

**Comment by jsarha** [comment]
>

**Comment by jsarha** [comment]
> @jsarha, thanks for the fix, it works great!
> 
> It would be really great to understand _why_ we need non blocking DMA and how that actually works for the probes and why it is bad for the dtrace.
> 

I am not sure about the reason why blocking DMA does not work with probes, but the reasoning behind getting rid of it in dtrace was a speculation that the IPC telling about the progress of the dma buffer may reach host before the buffer transfer is completed.

I do not thing there is such a problem with probes where the task doing to copy is scheduled periodically and the whole transfer is synchronous. I wonder if the problem is that the previous transfer is not complete and the work function has hot exit before the task is scheduled again?

**Comment by jsarha** [comment]
Hmm. Looks like I accidentally closed the PR. Reopened it.

**Comment by lgirdwood** [comment]
@jsarha @ujfalusi lets focus on getting this working with Zephyr. No point wasting effort with xtos related questions.

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**